### PR TITLE
Include folly as a submodule

### PR DESCRIPTION
**Pull request by hmusta**
An alternative method for using `folly::fbvector` and `folly::small_vector` without needing to install Folly.

### CODE REVIEW

**Comment by karasikov** [comment]
folly also has jemalloc, which we also want to use.
Not sure it makes sense to include it as a submodule.
I think an optional dependency (as before) is better, actually.

How does this make it easier to build?

**Comment by hmusta** [comment]
A lot of people have trouble installing folly, especially on Linux systems.

Regarding jemalloc: good point. We can very easily restore that my having `jemalloc` as a prerequisite and passing `-ljemalloc` to the linker (which was previously provided by the folly static lib)

**Comment by karasikov** [comment]
I think it's easier to install everything in one package rather than requiring more prerequisites.
If we add jemalloc, glog, etc., than it's gonna be equally hard to build, but in that case it will not be optional anymore.

If you think you can solve all those issues with linux, then go ahead.
But I think eventually you either remove some functionalities of folly or won't achieve any simplification.

Currently, folly is not required and one can easily compile metagraph without it.

Regarding compiling folly on linux, what issues does it have?

**Comment by hmusta** [comment]
> I think it's easier to install everything in one package rather than requiring more prerequisites.
> If we add jemalloc, glog, etc., than it's gonna be equally hard to build, but in that case it will not be optional anymore.
We can fix this by adding a check in the `CMakeLists.txt`, so the folly vectors are only used if these libraries are already present. glog and jemalloc are readily available as standard packages in Linux distros, whereas folly isn't.

> 
> If you think you can solve all those issues with linux, then go ahead.
> But I think eventually you either remove some functionalities of folly or won't achieve any simplification.
> 
> Currently, folly is not required and one can easily compile metagraph without it.
> 
> Regarding compiling folly on linux, what issues does it have?
Since it's a library that's neither standard, nor widely used, it requires a lot of extra steps to be taken to set up.

File: README.md [hunk 1]
```diff
@@ -6,23 +6,20 @@
...
-- folly (optional)
+- boost, jemalloc, gperftools (optional)
 - Python 3 (for running integration tests)
 
-All can be installed with [brew](https://brew.sh) or [linuxbrew](https://linuxbrew.sh) (does not require root)
-
 For compiling with **AppleClang**, the prerequisites can be installed as easy as:
 ```
-brew install libomp cmake make htslib boost folly
+brew install libomp cmake make htslib jemalloc gperftools
```
Comment (karasikov) [inline README.md:14]: what about boost, is it not needed anymore? But it's in the list of prerequisites.

File: metagraph/CMakeListsKMC.txt.in [hunk 2]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.2)
+cmake_minimum_required(VERSION 2.8.12)
```
Comment (karasikov) [inline metagraph/CMakeListsKMC.txt.in:1]: does CMakeListsKMC.txt.in not work with 2.8.2?

File: metagraph/CMakeListsHelpers.txt [hunk 3]
```diff
@@ -72,44 +72,57 @@ endfunction()
...
   )
-  if(LIBERTY)
-    set(FOLLY_LIBRARIES ${FOLLY_LIBRARIES} ${LIBERTY})
-  endif()
 else()
   set(FOLLY_LIBRARIES "")
-  message(WARNING "Folly (Facebook Open Source Library) not found."
-  " Install folly to reduce memory taken up by the Row-Compressed annotator"
+  message(WARNING "Prerequisites for Folly (Facebook Open Source Library) not found."
+  " Install Boost, jemalloc, and gperftools to reduce memory taken up by the Row-Compressed annotator"
```
Comment (karasikov) [inline metagraph/CMakeListsHelpers.txt:125]:
```suggestion
  " Install Boost, jemalloc, and gperftools to compile with folly and reduce memory taken up by the Row-Compressed annotator"
```

File: metagraph/CMakeListsHelpers.txt [hunk 4]
```diff
@@ -72,44 +72,57 @@ endfunction()
...
+  message(WARNING "gperftools not found")
+endif()
+
+if(jemalloc_FOUND AND tcmalloc_FOUND AND Boost_FOUND)
+message(STATUS "Configuring Folly")
+  string(APPEND CMAKE_CXX_FLAGS " -D_USE_FOLLY -DUSE_JEMALLOC -DUSE_TCMALLOC")
+  include(external-libraries/folly/CMake/FollyConfigChecks.cmake)
+  configure_file(
+    external-libraries/folly/CMake/folly-config.h.cmake
+    ${CMAKE_BINARY_DIR}/external-libraries/folly/include/folly/folly-config.h
```
Comment (karasikov) [inline metagraph/CMakeListsHelpers.txt:106]: can we move all this this to another cmake file find_folly, similarly to CMakeListsKMC.txt.in?

**Review by karasikov — CHANGES_REQUESTED** [review]

File: metagraph/CMakeListsKMC.txt.in (hunk 2 above)
Comment (hmusta) [reply metagraph/CMakeListsKMC.txt.in:1]: It was producing warnings that support for version <2.8.12 is deprecated and going to be phased out, so I updated this to suppress those.

File: README.md (hunk 1 above)
Comment (hmusta) [reply README.md]: Sorry, I had made this change and forgot to commit it.

File: .gitmodules [hunk 5]
```diff
@@ -62,3 +62,6 @@
 [submodule "metagraph/external-libraries/Simple-Web-Server"]
 	path = metagraph/external-libraries/Simple-Web-Server
 	url = https://gitlab.com/eidheim/Simple-Web-Server.git
+[submodule "metagraph/external-libraries/folly"]
+	path = metagraph/external-libraries/folly
+	url = https://github.com/facebook/folly
```
Comment (karasikov) [inline .gitmodules:67]: Since it's optional, maybe we should download it in cmake, like googletests?

File: README.md [hunk 6]
```diff
@@ -3,26 +3,35 @@
...
 
 ### Prerequisites
-- cmake 3.10
+- cmake 3.10 or higher
 - GNU GCC with C++17 (gcc-8 or higher), LLVM Clang (clang-7 or higher), or AppleClang (clang-1100.0.33.8 or higher)
 - HTSlib
-- folly (optional)
-- Python 3 (for running integration tests)
 
-All can be installed with [brew](https://brew.sh) or [linuxbrew](https://linuxbrew.sh) (does not require root)
```
Comment (karasikov) [inline README.md:12]: I think it's better to keep this line (maybe just rephrase this a bit), it shows how to install the dependencies without root.

**Review by karasikov — CHANGES_REQUESTED** [review]

File: .gitmodules (hunk 5 above)
Comment (hmusta) [reply .gitmodules:67]: googletests is a submodule, so it's downloaded every time. also, Folly only takes up 19MB, so it's not that big of a download.

I think making this change will add unnecessary complexity.

**Review by karasikov — APPROVED** [review]

File: metagraph/CMakeLists.txt [hunk 7]
```diff
@@ -364,10 +337,13 @@ add_library(metagraph-cli STATIC ${metagraph_cli_files})
...
 set_target_properties(metagraph PROPERTIES OUTPUT_NAME "metagraph_${CMAKE_DBG_ALPHABET}")
 
+find_package(BZip2 REQUIRED)
+
 set(METALIBS ${METALIBS}
   -lKMC
-  -lhts -lz -lbz2
+  -lhts -lz
   -lsdsl
+  ${BZIP2_LIBRARIES}
```
Comment (karasikov) [inline metagraph/CMakeLists.txt:346]: It's a dependency of HTSlib, right?
```suggestion
  -lhts -lz ${BZIP2_LIBRARIES}
  -lsdsl
```

**Review by karasikov — CHANGES_REQUESTED** [review]

File: metagraph/CMakeLists.txt (hunk 7 above)
Comment (hmusta) [reply metagraph/CMakeLists.txt:346]: I didn't notice that, actually. I added this because it's a requirement for KMC.

Comment (karasikov) [reply metagraph/CMakeLists.txt]: Then `-lKMC ${BZIP2_LIBRARIES}`... to make it clear why we include that.

**Review by karasikov — APPROVED** [review]

**Pull request merged**